Suggest `i += 1` when we see `i++` or `++i`

camelid opened the pull request:
Closes #83502 (for `i++` and `++i`; `--i` should be covered by #82987, and `i--`
is tricky to handle).

This is a continuation of #83536.

r? @estebank

camelid committed:
Suggest `i += 1` when we see `i++` or `++i`

camelid commented on src/test/ui/parser/increment.stderr:
```diff
@@ -0,0 +1,52 @@
+   = note: Rust has no dedicated increment operator
+help: try using `+= 1` instead
+   |
+LL |     while i += 1 < 5 {
```
I'm not sure how to fix this suggestion so it works like the pre-increment suggestion. The issue is that by the time we get to the second `+`, we've forgotten about the `i` token, so I don't see a good way of getting the starting span.

camelid replied:
Note to self: If we fix this, I should make sure to add `// run-rustfix` to the test.

camelid replied:
Fixed in newer version (still working on it, so haven't pushed).

camelid commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -1788,6 +1788,49 @@ impl<'a> Parser<'a> {
+
+            err.note("Rust has no dedicated increment operator");
+
+            if self.prev_token.kind == TokenKind::Semi {
```
This seems to be a good heuristic so that we don't suggest `{ i += 1; i};`. However, I wonder if there are places where this heuristic yields a false positive, and we emit code that is different from what the user meant. It seems unlikely, but should we change the applicability for this version of the suggestion to `MaybeIncorrect` just in case?

estebank replied:
Looking at all these changes, I feel like a better place to put this is in `parse_lit_expr`. That way we can

* give a specific error instead of the generic "expected expression":

```
error: Rust has no postfix increment operator
  --> $DIR/increment.rs:3:7
   |
LL |     i++;
   |      ^^ not a valid unary operator
   |
help: use in-place increment instead
   |
LL |     i += 1;
   |       ~~~~
```

* we can recover the parser and return that as an expression (which would also give you mutability and type errors if appropriate)
* gives us a good place where we can look for the prefix version: you add a `None if matches!(self.token.kind, Plus | Minus)` (and a lookahead 1 and 2 to check for another `Plus | Minus` and for a numeric literal or ident) case in the match in that function
* you can still check for whether the next thing will be a `Semi` to not give the `{}` surrounded version

`parse_lit_expr` is already the fall-through of `parse_bottom_expr`, which is effectively where expressions are parsed.

camelid replied:
> you can still check for whether the next thing will be a `Semi` to not give the `{}` surrounded version

Actually I need to check for whether the *previous* thing was a `Semi`, because otherwise on code like this:

```rust
let y = x++;

we will suggest this:

```rust
let y = x += 1;
```

How should I get around this? Will the `prev_token` in `parse_lit_expr` work?

camelid replied:
It looks like by the time `parse_lit_expr` is invoked for the postfix version, we've lost the context of the previous `;` because the `prev_token` is the variable being incremented. What should I do?

camelid replied:
Actually, it looks like `prev_token` is the first `+`, not even the variable. This seems like a problem...

camelid replied:
I moved the code to `parse_assoc_expr_with()` (and extracted a bunch of helper functions), which has more information.

b-naber commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -1788,6 +1788,49 @@ impl<'a> Parser<'a> {
             self.sess.expr_parentheses_needed(&mut err, *sp);
         }
         err.span_label(span, "expected expression");
+        if self.prev_token.kind == TokenKind::BinOp(token::Plus)
```
Can you maybe add a short comment here explaining what is done is this block, so that one can more easily skip reading this block?

estebank replied:
Or even better move it to its own method.

camelid replied:
It's in its own method (actually, multiple methods) now.

camelid committed:
Move increment checks to improve errors

camelid commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -1788,6 +1788,49 @@ impl<'a> Parser<'a> {
+            err.note("Rust has no dedicated increment operator");
+            err.span_suggestion_verbose(
+                span,
+                "try using `+= 1` instead",
```
The version you're suggesting sounds confusing to me – `i++` is also "in-place increment".

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3582065162) [(plain)](https://github.com/rust-lang/rust/commit/73917be700b1fb0f9e1677202a53280c8cffb9f6/checks/3582065162/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_parse v0.0.0 (/checkout/compiler/rustc_parse)
    Checking rustc_ast_lowering v0.0.0 (/checkout/compiler/rustc_ast_lowering)
    Checking chalk-engine v0.55.0
    Checking rustc_middle v0.0.0 (/checkout/compiler/rustc_middle)
error: variant is never constructed: `Dec`
    |
156 |     Dec,
    |     ^^^
    |
    |
    = note: `-D dead-code` implied by `-D warnings`
error: could not compile `rustc_parse` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
Build completed unsuccessfully in 0:02:15
```

</details>

camelid commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -1076,6 +1120,145 @@ impl<'a> Parser<'a> {
+            Applicability::MachineApplicable,
+        );
+        err.emit();
+        // TODO: recover
```
It seems like recovering from this might be fairly involved. Is it worth the added code complexity to return an expression like `{ i += 1; i }` instead of having the user just fix their code and rebuild to see further errors?

davidtwco replied:
At least initially, if recovery is challenging to implement, we can skip that so that this can land.

camelid replied:
Sounds good.

camelid commented on compiler/rustc_parse/src/parser/expr.rs:
```diff
@@ -230,6 +230,18 @@ impl<'a> Parser<'a> {
+                let op_span = self.prev_token.span.to(self.token.span);
+                // Eat the second `+`
+                self.bump();
+                // TODO: implement
+                let start_is_semi = false;
```
I'm not sure how to figure out if the previous token is a semicolon. It seems like even at the start of this function, `prev_token` refers to the `i`.

davidtwco replied:
I can't see a way to do this in this function outside of threading some context through (which we probably don't want?) or finding somewhere else to hook this diagnostic in at.

camelid replied:
Any ideas what I should do?

davidtwco replied:
I think it would be best if just assume there isn't a semicolon here and downgrade the applicability of the suggestion accordingly (perhaps adjusting the message to emphasise that the user _might_ want the suggestion). Then we can land this as an improvement and revisit that case in future.

camelid replied:
Ok, I think I found a good workaround. Basically, emit both suggestions. However, rustfix is crashing:

```

---- [ui] ui/parser/increment-autofix.rs stdout ----
thread '[ui] ui/parser/increment-autofix.rs' panicked at 'failed to apply suggestions for "/Users/noahlev/code/rust/r1/src/test/ui/parser/increment-autofix.rs" with rustfix: Cannot replace slice of data that was already replaced', src/tools/compiletest/src/runtest.rs:3131:17
```

Any ideas?

davidtwco replied:
Can you use [`span_suggestions`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_errors/diagnostic/struct.Diagnostic.html#method.span_suggestions)?

camelid replied:
Ah, is that new? The only problem with it is I can't have different messages, which might be confusing.

davidtwco replied:
I think it's the only option we've got, so we'll just need to find a message that works.

camelid commented on src/test/ui/parser/increment.stderr:
```diff
@@ -0,0 +1,47 @@
+LL |     i++;
+   |      ^^ not a valid postfix operator
+   |
+help: use `+= 1` instead
+   |
+LL |     { let tmp = i; i += 1; tmp };
+   |     +++++++++++  ~~~~~~~~~~~~~~~
```
This is because of the code I commented on above where I don't know how to tell if this previous token is a semicolon. I'm not even certain if that's a good heuristic, because if we're at the beginning of a function, the previous token will be a `{`. What I really want is to know if are at the beginning of a new statement. Is there any way to determine that?

davidtwco replied:
I don't think there's a great way of knowing that, outside of keeping track of more context - it can be left for a follow-up and we can consider the suggestion possibly correct.

camelid replied:
Well, it is actually "correct", just unnecessarily complicated.

camelid replied:
See https://github.com/rust-lang/rust/pull/88672#discussion_r809538489.

camelid commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -1076,6 +1120,145 @@ impl<'a> Parser<'a> {
+        err.multipart_suggestion(
+            &format!("use `{}= 1` instead", kind.op.chr()),
+            vec![
+                (pre_span, "{ let tmp = ".to_string()),
+                (post_span, format!("; {} {}= 1; tmp }}", ident, kind.op.chr())),
```
One problem here is if the user's code was `tmp++`. Then the generated code will be incorrect. Is there a better way to do this (especially because it's currently marked as machine-applicable)?

camelid replied:
Added a workaround.

camelid commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -1076,6 +1120,145 @@ impl<'a> Parser<'a> {
+        );
+        err.span_label(op_span, &format!("not a valid {} operator", kind.fixity));
+
+        if let ExprKind::Path(_, path) = &base.kind {
+            if let [segment] = path.segments.as_slice() {
+                let ident = segment.ident;
```
This doesn't work for code like `foo.bar++`. Instead, we fall back to the base case:

```
error: Rust has no postfix increment operator
 --> bar.rs:2:12
  |
2 |     foo.bar++;
  |            ^^ not a valid postfix operator
  |
  = help: use `+= 1` instead
```

Is there a way to fix this simply?

davidtwco replied:
Just guessing from looking at the AST we generate on Compiler Explorer, I assume you don't just need to peel back N layers of `ExprKind::Field` to get to the innermost `ExprKind::Path` and then continue as normal?

camelid replied:
Looking at this now, I realize this code isn't the issue (we can just use `span_to_snippet` to avoid the AST manipulation). The issue is that when parsing `++foo.bar`, we need to determine how far ahead to parse. Currently it just parses a single path segment, but that doesn't work for `foo.bar`. Thoughts?

camelid commented:
Is there any way this could accidentally break macros that have `i++`–style syntax? I tested it a bit, and I didn't find any issues, but I thought it'd be worth bringing up.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3582122170) [(plain)](https://github.com/rust-lang/rust/commit/f72a70170de71d5043e12223ccd95c3fee9b508f/checks/3582122170/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
tidy: Skipping binary file check, read-only filesystem
Checking which error codes lack tests...
* 627 error codes
* highest error code: E0785
tidy error: /checkout/compiler/rustc_parse/src/parser/diagnostics.rs:1221: TODO is deprecated; use FIXME
tidy error: /checkout/compiler/rustc_parse/src/parser/diagnostics.rs:1242: TODO is deprecated; use FIXME
tidy error: /checkout/compiler/rustc_parse/src/parser/diagnostics.rs:1260: TODO is deprecated; use FIXME
tidy error: /checkout/compiler/rustc_parse/src/parser/expr.rs:244: TODO is deprecated; use FIXME
Found 0 error codes with no tests
Done!
* 350 features
some tidy checks failed
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "16"


Build completed unsuccessfully in 0:00:12
```

</details>

camelid commented:
(CI is just failing because I used `// TODO`, which was intentional so they would be resolved before this is merged.)

pnkfelix commented:
Hi @camelid ; we visited this for triage. 

It looks like you had some questions you wanted answered, but also, @estebank had a suggestion for a [change](https://github.com/rust-lang/rust/pull/88672#discussion_r702776614) they wanted to see.

What is the status on those fronts? Do you think you are stuck waiting for more feedback from @estebank, or can you make forward progress on the suggestion they provided in the meantime?

camelid commented:
It's been a while, but my recollection is that I implemented most of the user-facing side of what estebank suggested, but in a different way in the parser. The way the change is implemented currently works pretty well, though sometimes the suggestions aren't as good as they could be. Perhaps it would be good to accept the current state of output as good enough for an "MVP PR" since it can be improved further in the future. That's of course assuming there aren't any major issues in the output.

So, in summary: I think I need estebank to see if the impl looks okay and the error output looks good enough for a first implementation.

camelid commented:
@estebank could you take another look at this when you get a chance?

bors commented:
:umbrella: The latest upstream changes (presumably #91933) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

camelid commented:
I'll wait to rebase until this is reviewed in case it needs more rebases.

camelid commented:
r? @davidtwco

davidtwco commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -1076,6 +1122,145 @@ impl<'a> Parser<'a> {
+                    UnaryFixity::Post => (ident.span.shrink_to_lo(), op_span),
+                };
+
+                if !ident.is_reserved() {
```
Is there any way we could cut down on how nested this expression is getting?

camelid replied:
I added early returns. How is it now?

davidtwco reviewed (commented):
Apologies for how long it has taken for this to be reviewed again, this looks  basically good to me (at least as a starting point for this suggestion), I've left some comments on your open questions.

camelid committed:
Remove error recovery todos

camelid committed:
Reduce rightward drift

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5211050620?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/2f5756d776d54553c24b7a2ac7f644822dc8d380/checks/5211050620/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
* highest error code: E0787
Found 504 error codes
Found 0 error(s) in error codes
Done!
tidy error: /checkout/compiler/rustc_parse/src/parser/expr.rs:276: TODO is deprecated; use FIXME
some tidy checks failed
Build completed unsuccessfully in 0:00:12
```

</details>

davidtwco commented:
I think we should prioritize tidying up the suggestions this adds which do work so we can land those, and we can revisit the difficult cases in another PR.

camelid committed:
Refactor, handle fields better, add field tests

camelid committed:
Emit structured suggestions for field accesses too

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5239830023?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/1fbf8d0322a7647df278928eaa02b51f6fee5cf6/checks/5239830023/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
* highest error code: E0787
Found 504 error codes
Found 0 error(s) in error codes
Done!
tidy error: /checkout/compiler/rustc_parse/src/parser/expr.rs:276: TODO is deprecated; use FIXME
some tidy checks failed
Build completed unsuccessfully in 0:00:10
```

</details>

camelid committed:
Emit both subexp and standalone sugg for postfix

This solves the TODO.

camelid committed:
Improve handling of `tmp` variable name conflicts

camelid commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -1112,6 +1165,150 @@ impl<'a> Parser<'a> {
+    ) {
+        let mut msg = format!("use `{}= 1` instead", kind.op.chr());
+        if base_src.trim() == "tmp" {
+            msg.push_str(" (rename `tmp` so it doesn't conflict with your variable)");
```
Wording could probably be improved here.

davidtwco replied:
If we can detect this, why can't we just change the variable name in our suggestion?

camelid replied:
I thought about doing that but worried it might be overkill. My idea was to have a loop where we add underscores until the name doesn't match. Alternatively, it could append a number. Personally I prefer the underscore approach because I doubt it'll require more than one or two underscores, whereas `tmp4` could be a real variable.

camelid replied:
I just realized that there's no need for a loop because we're only comparing against one variable 🤦

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5240276320?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/a69f5388a48ed3b9f60d493e1f2040b36081b61f/checks/5240276320/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [ui] ui/associated-types/issue-36499.rs stdout ----
diff of stderr:

- error: leading `+` is not supported
-   --> $DIR/issue-36499.rs:4:9
+ error: Rust has no postfix increment operator
3    |
4 LL |     2 + +2;
-    |         ^ unexpected `+`
+    |       ^^^ not a valid postfix operator
+    |       ^^^ not a valid postfix operator
6    |
- help: try removing the `+`
+ help: use `+= 1` instead
8    |
+ LL |     { let tmp = 2 ; 2 += 1; tmp }2;
+    |     +++++++++++   ~~~~~~~~~~~~~~~
+ help: or, if you don't need to use it as an expression, change it to this
9 LL -     2 + +2;
- LL +     2 + 2;
- LL +     2 + 2;
+ LL +     2  += 12;
12 
13 error: aborting due to previous error


---
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/associated-types/issue-36499.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-types/issue-36499" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-types/issue-36499/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error: Rust has no postfix increment operator
  --> /checkout/src/test/ui/associated-types/issue-36499.rs:4:7
   |
LL |     2 + +2;
   |       ^^^ not a valid postfix operator
   |
help: use `+= 1` instead
   |
LL |     { let tmp = 2 ; 2 += 1; tmp }2;
   |     +++++++++++   ~~~~~~~~~~~~~~~
help: or, if you don't need to use it as an expression, change it to this
LL -     2 + +2;
LL +     2  += 12;
   | 


error: aborting due to previous error


------------------------------------------


---- [ui] ui/parser/increment-autofix.rs stdout ----
thread '[ui] ui/parser/increment-autofix.rs' panicked at 'failed to apply suggestions for "/checkout/src/test/ui/parser/increment-autofix.rs" with rustfix: Cannot replace slice of data that was already replaced', src/tools/compiletest/src/runtest.rs:3131:17
---- [ui] ui/parser/issues/issue-88276-unary-plus.rs stdout ----
diff of stderr:

10 LL +     let _ = 1;
10 LL +     let _ = 1;
11    | 
12 
- error: leading `+` is not supported
-   --> $DIR/issue-88276-unary-plus.rs:5:20
+ error: Rust has no postfix increment operator
15    |
15    |
16 LL |     let _ = (1.0 + +2.0) * +3.0;
-    |                    ^ unexpected `+`
+    |                  ^^^ not a valid postfix operator
- help: try removing the `+`
- help: try removing the `+`
+ help: use `+= 1` instead
20    |
+ LL |     let _ = ({ let tmp = 1.0 ; 1.0 += 1; tmp }2.0) * +3.0;
+    |              +++++++++++     ~~~~~~~~~~~~~~~~~
+ help: or, if you don't need to use it as an expression, change it to this
+    |
21 LL -     let _ = (1.0 + +2.0) * +3.0;
- LL +     let _ = (1.0 + 2.0) * +3.0;
+ LL +     let _ = (1.0  += 12.0) * +3.0;
24 
24 
25 error: leading `+` is not supported

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/parser/issues/issue-88276-unary-plus/issue-88276-unary-plus.stderr
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/parser/issues/issue-88276-unary-plus/issue-88276-unary-plus.stderr
thread '[ui] ui/parser/issues/issue-88276-unary-plus.rs' panicked at 'failed to apply suggestions for "/checkout/src/test/ui/parser/issues/issue-88276-unary-plus.rs" with rustfix: Cannot replace slice of data that was already replaced', src/tools/compiletest/src/runtest.rs:3131:17

failures:
    [ui] ui/associated-types/issue-36499.rs
    [ui] ui/parser/increment-autofix.rs
```

</details>

camelid committed:
Bless tests

camelid committed:
Improve function names

camelid committed:
Add heuristic to avoid treating `x + +2` as increment

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5241110795?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/4af5db80ae3219ec04c140944a439a8e9c4022eb/checks/5241110795/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.........................................iii........................................................ 12600/12646
..............................................
failures:

---- [ui] ui/parser/increment-autofix.rs stdout ----
thread '[ui] ui/parser/increment-autofix.rs' panicked at 'failed to apply suggestions for "/checkout/src/test/ui/parser/increment-autofix.rs" with rustfix: Cannot replace slice of data that was already replaced', src/tools/compiletest/src/runtest.rs:3131:17

failures:
    [ui] ui/parser/increment-autofix.rs

```

</details>

davidtwco commented on src/test/ui/parser/increment-autofix.stderr:
```diff
@@ -0,0 +1,89 @@
+   |
+LL |     { let tmp = i; i += 1; tmp };
+   |     +++++++++++  ~~~~~~~~~~~~~~~
+help: or, if you don't need to use it as an expression, change it to this
```
```suggestion
help: or, if you don't need to use the incremented value
```
I think this might be a little bit clearer, avoids terminology like "expression" which might not be obvious to a beginner.

camelid replied:
Yeah, I was a bit concerned about saying "expression". However, "the incremented value" is actually not quite right because the suggestion yields the pre-increment value. What about this?

```suggestion
help: or, if you don't need to use the pre-increment value
```

or this?

```suggestion
help: or, if you don't need to use the old value
```

davidtwco replied:
Either of these are good, I don't mind which you pick, I think the first is clearer but it also has more terminology

davidtwco commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -150,6 +150,59 @@ impl AttemptLocalParseRecovery {
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum IncOrDec {
+    Inc,
+    // FIXME: `i--` recovery isn't implemented yet
```
Are you intending to do this in the current PR?

camelid replied:
No, `i--` and `--i` are tricky to handle because in many cases they're valid code. I'd rather save them for later so this PR can land.

davidtwco commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -150,6 +150,59 @@ impl AttemptLocalParseRecovery {
+    /// Is this increment/decrement its own statement?
+    ///
+    /// This is `None` when we are unsure.
+    standalone: Option<bool>,
```
nit: it might be more self-documenting if this was an enum with three variants, I skipped over the comment at first and it took a second to understand what situation `None` corresponded to

camelid replied:
Done, good idea.

davidtwco reviewed (changes requested):
This is great, left a few more comments, should be able to land this soon.

camelid committed:
Make `standalone` an enum

camelid committed:
Change temporary variable name if it would conflict

camelid commented:
Thanks for the helpful reviews by the way! :)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5254875173?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/aefa04d6abeb5d1d0b8f6b66cd147b8b5aa4d919/checks/5254875173/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_query_system v0.0.0 (/checkout/compiler/rustc_query_system)
    Checking rustc_parse v0.0.0 (/checkout/compiler/rustc_parse)
    Checking rustc_middle v0.0.0 (/checkout/compiler/rustc_middle)
    Checking rustc_ast_lowering v0.0.0 (/checkout/compiler/rustc_ast_lowering)
error[E0636]: the feature `let_else` has already been declared
  |
  |
8 | #![feature(let_else)]

For more information about this error, try `rustc --explain E0636`.
error: could not compile `rustc_parse` due to previous error
warning: build failed, waiting for other jobs to finish...
```

</details>

camelid commented:
`let_else` was added a couple days ago so I had to rebase to remove my copy of it.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5254923221?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/8f6f1e9ca01f1094d331dce2ca1d0e5d00228931/checks/5254923221/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
...........................................iii...................................................... 12600/12648
................................................
failures:

---- [ui] ui/parser/increment-autofix.rs stdout ----
thread '[ui] ui/parser/increment-autofix.rs' panicked at 'failed to apply suggestions for "/checkout/src/test/ui/parser/increment-autofix.rs" with rustfix: Cannot replace slice of data that was already replaced', src/tools/compiletest/src/runtest.rs:3131:17

failures:
    [ui] ui/parser/increment-autofix.rs

```

</details>

camelid committed:
Use `multipart_suggestions`

This records that the suggestions are mutually-exclusive (i.e., only one
should be applied).

camelid commented:
Hmm, rustfix is still panicking, even with now using `multipart_suggestions` (notice that it's plural).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5282541298?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/6a89e97ab14cd4c6be70b47fd139c87fd90150b7/checks/5282541298/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
......................................................iii........................................... 12600/12659
...........................................................
failures:

---- [ui] ui/parser/increment-autofix.rs stdout ----
thread '[ui] ui/parser/increment-autofix.rs' panicked at 'failed to apply suggestions for "/checkout/src/test/ui/parser/increment-autofix.rs" with rustfix: Cannot replace slice of data that was already replaced', src/tools/compiletest/src/runtest.rs:3142:17

failures:
    [ui] ui/parser/increment-autofix.rs

```

</details>

davidtwco commented:
> Hmm, rustfix is still panicking, even with now using `multipart_suggestions` (notice that it's plural). How do I fix this?

Apologies for the delay in responding - I'm not entirely sure, could you try commenting some of the suggestions locally and see if removing any of those fixes the problem? Perhaps the `IsStandalone::Maybe` case? Failing that, I'd see if you can't work out what is going on within `rustfix` to cause it to fail, as much as that's a bunch more work.

camelid commented:
I opened an issue for rustfix: https://github.com/rust-lang/rustfix/issues/209. I'm pretty sure the issue is on its end, so I'm not sure we can fix it in rustc.

How about if we mark the suggestions as non-MachineApplicable in the meantime until the bug is resolved?

bors commented:
:umbrella: The latest upstream changes (presumably #95056) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

davidtwco commented:
> I opened an issue for rustfix: [rust-lang/rustfix#209](https://github.com/rust-lang/rustfix/issues/209). I'm pretty sure the issue is on its end, so I'm not sure we can fix it in rustc.
> 
> How about if we mark the suggestions as non-MachineApplicable in the meantime until the bug is resolved?

I think that makes sense.

camelid commented:
Turns out the panic is still happening without MachineApplicable. I asked ehuss about it on the rustfix issue.

camelid committed:
Fix rustfix panic on test

`run-rustfix` applies all suggestions regardless of their Applicability.
There's a flag, `rustfix-only-machine-applicable`, that does what it
says, but then the produced `.fixed` file would have invalid code from
the suggestions that weren't applied. So, I moved the cases of postfix
increment, in which case multiple suggestions are given, to the
`-notfixed` test, which does not run rustfix.

I also changed the Applicability to Unspecified since MaybeIncorrect
requires that the code be valid, even if it's incorrect.

camelid commented:
Ok, I got the tests working now! (You can read the rustfix issue for the background on what the problem and solution were.) Thanks to ehuss for their help with solving this.

I'm going to rebase now and then this should finally be ready for merge!

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5669323605?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/c0bcea017914034d2aec9fb88662d8fd46f9e634/checks/5669323605/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_session v0.0.0 (/checkout/compiler/rustc_session)
    Checking rustc_attr v0.0.0 (/checkout/compiler/rustc_attr)
    Checking rustc_query_system v0.0.0 (/checkout/compiler/rustc_query_system)
    Checking rustc_parse v0.0.0 (/checkout/compiler/rustc_parse)
error[E0107]: this struct takes 1 generic argument but 0 generic arguments were supplied
    |
    |
228 |     fn emit(self, err: &mut DiagnosticBuilder<'_>) {
    |                             ^^^^^^^^^^^^^^^^^ expected 1 generic argument
    |
note: struct defined here, with 1 generic parameter: `G`
    |
    |
19  | pub struct DiagnosticBuilder<'a, G: EmissionGuarantee> {
help: add missing generic argument
    |
    |
228 |     fn emit(self, err: &mut DiagnosticBuilder<'_, G>) {


error[E0107]: this struct takes 1 generic argument but 0 generic arguments were supplied
    |
234 |         err: &mut DiagnosticBuilder<'_>,
    |                   ^^^^^^^^^^^^^^^^^ expected 1 generic argument
    |
    |
note: struct defined here, with 1 generic parameter: `G`
    |
    |
19  | pub struct DiagnosticBuilder<'a, G: EmissionGuarantee> {
help: add missing generic argument
    |
    |
234 |         err: &mut DiagnosticBuilder<'_, G>,

For more information about this error, try `rustc --explain E0107`.
error: could not compile `rustc_parse` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
```

</details>

camelid committed:
Fix from rebase

I changed the test functions to be `pub` rather than called from a
`main` function too, for easier future modification of tests.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5670340232?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/af8c4fe5b3f036e887bd49593da8bceaa138fdaf/checks/5670340232/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
............................................................i.i..................................... 8300/12764
...............................................iiii................................................. 8400/12764
....................................................................i............................... 8500/12764
........i...................................................................i....................... 8600/12764
................................................................F.F................................. 8700/12764
.................................................................................................... 8900/12764
............i....................................................................................... 9000/12764
.................................................................................................... 9100/12764
.................................................................................................... 9200/12764
---
...........................................................iii...................................... 12700/12764
................................................................
failures:

---- [ui] ui/parser/increment-autofix.rs stdout ----

14   --> $DIR/increment-autofix.rs:11:11
15    |
15    |
16 LL |     while ++i < 5 {
-    |           ^^ not a valid prefix operator
+    |     ----- ^^ not a valid prefix operator
+    |     |
+    |     while parsing the condition of this `while` expression
18    |
19 help: use `+= 1` instead

37   --> $DIR/increment-autofix.rs:25:11
38    |
38    |
39 LL |     while ++tmp < 5 {
-    |           ^^ not a valid prefix operator
+    |     ----- ^^ not a valid prefix operator
+    |     |
+    |     while parsing the condition of this `while` expression
41    |
42 help: use `+= 1` instead


The actual stderr differed from the expected stderr.
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/parser/increment-autofix/increment-autofix.stderr
To only update this specific test, also pass `--test-args parser/increment-autofix.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/parser/increment-autofix.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/parser/increment-autofix" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/parser/increment-autofix/auxiliary"
stdout: none
--- stderr -------------------------------
error: Rust has no prefix increment operator
  --> /checkout/src/test/ui/parser/increment-autofix.rs:5:5
   |
LL |     ++i; //~ ERROR Rust has no prefix increment operator
   |     ^^ not a valid prefix operator
   |
help: use `+= 1` instead
   |
LL -     ++i; //~ ERROR Rust has no prefix increment operator
LL +     i += 1; //~ ERROR Rust has no prefix increment operator

error: Rust has no prefix increment operator
  --> /checkout/src/test/ui/parser/increment-autofix.rs:11:11
   |
   |
LL |     while ++i < 5 {
   |     ----- ^^ not a valid prefix operator
   |     |
   |     while parsing the condition of this `while` expression
   |
help: use `+= 1` instead
   |
LL |     while { i += 1; i } < 5 {
   |           ~   +++++++++
error: Rust has no prefix increment operator
  --> /checkout/src/test/ui/parser/increment-autofix.rs:19:5
   |
   |
LL |     ++tmp; //~ ERROR Rust has no prefix increment operator
   |     ^^ not a valid prefix operator
   |
help: use `+= 1` instead
   |
LL -     ++tmp; //~ ERROR Rust has no prefix increment operator
LL +     tmp += 1; //~ ERROR Rust has no prefix increment operator

error: Rust has no prefix increment operator
  --> /checkout/src/test/ui/parser/increment-autofix.rs:25:11
   |
   |
LL |     while ++tmp < 5 {
   |     ----- ^^ not a valid prefix operator
   |     |
   |     while parsing the condition of this `while` expression
   |
help: use `+= 1` instead
   |
LL |     while { tmp += 1; tmp } < 5 {

error: aborting due to 4 previous errors
------------------------------------------



---- [ui] ui/parser/increment-notfixed.rs stdout ----

16   --> $DIR/increment-notfixed.rs:17:12
17    |
17    |
18 LL |     while i++ < 5 {
-    |            ^^ not a valid postfix operator
+    |     -----  ^^ not a valid postfix operator
+    |     |
+    |     while parsing the condition of this `while` expression
20    |
21 help: use `+= 1` instead

44   --> $DIR/increment-notfixed.rs:31:14
45    |
45    |
46 LL |     while tmp++ < 5 {
-    |              ^^ not a valid postfix operator
+    |     -----    ^^ not a valid postfix operator
+    |     |
+    |     while parsing the condition of this `while` expression
48    |
49 help: use `+= 1` instead


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/parser/increment-notfixed/increment-notfixed.stderr
To only update this specific test, also pass `--test-args parser/increment-notfixed.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/parser/increment-notfixed.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/parser/increment-notfixed" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/parser/increment-notfixed/auxiliary"
stdout: none
--- stderr -------------------------------
error: Rust has no postfix increment operator
  --> /checkout/src/test/ui/parser/increment-notfixed.rs:11:6
   |
LL |     i++; //~ ERROR Rust has no postfix increment operator
   |      ^^ not a valid postfix operator
   |
help: use `+= 1` instead
   |
LL |     { let tmp = i; i += 1; tmp }; //~ ERROR Rust has no postfix increment operator
   |     +++++++++++  ~~~~~~~~~~~~~~~
LL -     i++; //~ ERROR Rust has no postfix increment operator
LL +     i += 1; //~ ERROR Rust has no postfix increment operator

error: Rust has no postfix increment operator
  --> /checkout/src/test/ui/parser/increment-notfixed.rs:17:12
   |
   |
LL |     while i++ < 5 {
   |     -----  ^^ not a valid postfix operator
   |     |
   |     while parsing the condition of this `while` expression
   |
help: use `+= 1` instead
   |
LL |     while { let tmp = i; i += 1; tmp } < 5 {
   |           +++++++++++  ~~~~~~~~~~~~~~~
LL -     while i++ < 5 {
LL +     while i += 1 < 5 {

error: Rust has no postfix increment operator
  --> /checkout/src/test/ui/parser/increment-notfixed.rs:25:8
   |
   |
LL |     tmp++; //~ ERROR Rust has no postfix increment operator
   |        ^^ not a valid postfix operator
   |
help: use `+= 1` instead
   |
LL |     { let tmp_ = tmp; tmp += 1; tmp_ }; //~ ERROR Rust has no postfix increment operator
   |     ++++++++++++    ~~~~~~~~~~~~~~~~~~
LL -     tmp++; //~ ERROR Rust has no postfix increment operator
LL +     tmp += 1; //~ ERROR Rust has no postfix increment operator

error: Rust has no postfix increment operator
  --> /checkout/src/test/ui/parser/increment-notfixed.rs:31:14
   |
   |
LL |     while tmp++ < 5 {
   |     -----    ^^ not a valid postfix operator
   |     |
   |     while parsing the condition of this `while` expression
   |
help: use `+= 1` instead
   |
LL |     while { let tmp_ = tmp; tmp += 1; tmp_ } < 5 {
   |           ++++++++++++    ~~~~~~~~~~~~~~~~~~
LL -     while tmp++ < 5 {
LL +     while tmp += 1 < 5 {

error: Rust has no postfix increment operator
  --> /checkout/src/test/ui/parser/increment-notfixed.rs:39:16
   |
   |
LL |     foo.bar.qux++;
   |                ^^ not a valid postfix operator
   |
help: use `+= 1` instead
   |
LL |     { let tmp = foo.bar.qux; foo.bar.qux += 1; tmp };
   |     +++++++++++            ~~~~~~~~~~~~~~~~~~~~~~~~~
LL -     foo.bar.qux++;
LL +     foo.bar.qux += 1;

error: Rust has no postfix increment operator
  --> /checkout/src/test/ui/parser/increment-notfixed.rs:49:10
   |
   |
LL |     s.tmp++;
   |          ^^ not a valid postfix operator
   |
help: use `+= 1` instead
   |
LL |     { let tmp = s.tmp; s.tmp += 1; tmp };
   |     +++++++++++      ~~~~~~~~~~~~~~~~~~~
LL -     s.tmp++;
LL +     s.tmp += 1;

error: Rust has no prefix increment operator
  --> /checkout/src/test/ui/parser/increment-notfixed.rs:56:5
   |
   |
LL |     ++foo.bar.qux;
   |     ^^ not a valid prefix operator
   |
help: use `+= 1` instead
   |
LL -     ++foo.bar.qux;
LL +     foo.bar.qux += 1;

error: aborting due to 7 previous errors
------------------------------------------

```

</details>

camelid commented:
Ok, hopefully CI should pass now.

davidtwco commented:
This is great, thanks for seeing this through. Could you squash some of the commits? r=me after that.

camelid commented:
@davidtwco It's hard to squash without squashing all of them into one commit, and I'd rather keep the history to make it easier to understand the design choices for the code (e.g., when `git blame`ing). Is it okay to leave it unsquashed?

davidtwco commented:
@bors r+

bors commented:
:pushpin: Commit 4943688e9d0d0f34f2fda188e9a9d2777f90d186 has been approved by `davidtwco`

<!-- @bors r=davidtwco 4943688e9d0d0f34f2fda188e9a9d2777f90d186 -->
<!-- homu: {"type":"Approved","sha":"4943688e9d0d0f34f2fda188e9a9d2777f90d186","approver":"davidtwco"} -->

bors commented:
:hourglass: Testing commit 4943688e9d0d0f34f2fda188e9a9d2777f90d186 with merge 133859d680ffe31072b80f518377fa0c487359d9...
<!-- homu: {"type":"BuildStarted","head_sha":"4943688e9d0d0f34f2fda188e9a9d2777f90d186","merge_sha":"133859d680ffe31072b80f518377fa0c487359d9"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5804521242?check_suite_focus=true)
Approved by: davidtwco
Pushing 133859d680ffe31072b80f518377fa0c487359d9 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"davidtwco","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5804521242?check_suite_focus=true"},"merge_sha":"133859d680ffe31072b80f518377fa0c487359d9"} -->

rust-timer commented:
Finished benchmarking commit (133859d680ffe31072b80f518377fa0c487359d9): [comparison url](https://perf.rust-lang.org/compare.html?start=c1550e3f8c57b0ebb0edb3a7ea64727e668f669c&end=133859d680ffe31072b80f518377fa0c487359d9).

**Summary**: 
- Primary benchmarks: no relevant changes found
- Secondary benchmarks: no relevant changes found. 2 results were found to be statistically significant but too small to be relevant.


| | Regressions 😿 <br />(primary) | Regressions 😿 <br />(secondary) | Improvements 🎉 <br />(primary) | Improvements 🎉 <br />(secondary) | All 😿 🎉 <br />(primary) |
|:---:|:---:|:---:|:---:|:---:|:---:|
| count[^1] | 0 | 2 | 0 | 0 | 0 |
| mean[^2] | N/A | 1.4% | N/A | N/A | N/A |
| max | N/A | 2.6% | N/A | N/A | N/A |

[^1]: *number of relevant changes*
[^2]: *the arithmetic mean of the percent change*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression

camelid commented:
Thank you so much for helping me with this PR and being patient even though it took me forever to finish it! ❤️
